docs/builder/ref.md

```diff
@@ -0,0 +1,290 @@
+# Builder Library Reference
+
+## Top-Level Program Structure
+
+Here's the general structure of a program that uses the builder to generate Calyx code.
+
+```python
+# import the builder library
+import calyx.builder as cb
+
+
+# define my_subcomponent
+def add_my_subcomponent(prog):
+    # subcomponent definition here
+
+
+# define my_component
+def my_component(prog, my_subcomponent): 
+    # add the component to the program
+    my_component = prog.component("my_component")
+
+    # add my_subcomponent as a cell of my_component
+    my_subcomponent = my_component.cell("my_subcomponent", my_subcomponent)
+
+    # define a my_component group
+    with my_component.group("my_group") as my_group:
+      # group assignments here 
+
+  my_component.control += [my_group]
+
+
+# assemble the program
+def build():
+    prog = cb.Builder()
+    my_subcomponent = add_my_subcomponent(prog)
+    add_main(prog, my_subcomponent)
```

Comment:
Somewhere like this, you could explain the callee-caller relationship, which is a hair different from a component-subcomponent relationship

Reply:
I just mean that one component is saying that it might call another component later on, by instantiating it as a cell.

https://github.com/cucapra/calyx/blob/74802618ca0fb602172038ca231cf5435e0495ff/calyx-py/test/correctness/reduction_tree.py#LL140C17-L140C21

https://docs.calyxir.org/lang/multi-component.html?highlight=component#multi-component-designs

Reply:
FWIW I would just call this "instantiation" or "creating a cell" or something (as opposed to "calling").

Reply:
that is more how i was conceptualizing this! thank you!